packages/gestalt/src/Masonry.tsx

```diff
@@ -204,30 +204,43 @@ export default class Masonry<T> extends ReactComponent<Props<T>, State<T>> {
       props.measurementStore || Masonry.createMeasurementStore();
 
     this.positionStore = props.positionStore || Masonry.createMeasurementStore();
-    this.idxToDataMap = {};
 
     this.resizeObserver =
       /* eslint-disable-next-line no-underscore-dangle */
       props._dynamicHeights && typeof window !== 'undefined' && this.positionStore
         ? new ResizeObserver((entries) => {
+            let triggerUpdate = false;
             for (let i = 0; i < entries.length; i += 1) {
               const { target, contentRect } = entries[i];
               const idx = Number(target.getAttribute('data-grid-item-idx'));
 
               if (typeof idx === 'number') {
-                const item: T = this.idxToDataMap[idx];
-                recalcHeights({
-                  items: this.state.items,
-                  changedItem: item,
-                  newHeight: contentRect.height,
-                  positionStore: this.positionStore,
-                  measurementStore: this.state.measurementStore,
-                  /* eslint-disable-next-line no-underscore-dangle */
-                  getColumnSpanConfig: this.props._getColumnSpanConfig,
-                });
-                this.forceUpdate();
+                const item: T = this.state.items[idx];
+                const changedItemPosition = this.positionStore.get(item);
+                const newHeight = contentRect.height;
+
+                if (
+                  changedItemPosition &&
+                  newHeight > 0 &&
```

Comment:
Added this check just in case it could happen that a the resize is triggered for the item that is not mounted. This also fixed the vistualized issue but adding the layoutEffect also sounds good.

Reply:
instead of the if here, can we have `recalcHeight` return whether a change was applied?

```
const didUpdate = recalcHeights();
if (didUpdate) {
 this.forceUpdate()
}
```